### ISSUE TITLE
Quasiquote?

### ISSUE THREAD

**Issue by technomancy** [bakpakin/Fennel#47]
Once we get some more people writing macros it would probably be good to have quasiquote; this would allow us to remove the `list` and `sym` specials that are only visible inside `eval-compiler`.

**Comment by thatismatt** [comment]
I think we may be reaching this point... what do you think?

Does anyone have thoughts on how this should be implemented. e.g. new syntax (backtick, tilde, comma, at etc) vs "named" special forms (quasiquote, unquote, unquote-splicing etc).

**Comment by technomancy** [comment]
My inclination would be to start with the named versions and see how it goes; we can add punctuation versions later. Though do note that `unquote-splicing` is just `unpack`, so `quasiquote` and `unquote` should be enough.

**Comment by numberjay** [comment]
it seems to me that `(unquote-splicing x)` would really be `(unquote (unpack x))` and not just `(unpack x)`

but as `unpack` works only at the tail of an expression (if i'm not mistaken) than we still need the functionality of `unquote-splicing` which must not have this limitation, even if we end up writing it as a composition of `unquote` and `unpack` (good choice)

the named version is the way to go and the reader should take care of desugaring backticks, commas and the like to their corresponding named forms

the VERY first step, though, is implementing `quote` which I was not able to find in fennel and i'm not sure how to implement it if not as a primitive special form

other than that, `quasiquote` does not need any special treatment and can be implemented as a macro

**Comment by thatismatt** [comment]
Yes, you can do that, like this:
`my-macros.fnl`:
```
{:plus->minus
 (fn [x ...]
   (table.remove x 1)
   (table.insert x 1 (sym "-"))
   x)}
```
`main.fnl`:
```
(require-macros :my-macros)

(print (plus->minus (+ 2 1)))
```
Hope that helps :)

**Comment by thatismatt** [comment]
Basically the Fennel AST is just Lua tables, so you can use all the normal table functions as you would with Lua data.

**Comment by bb010g** [comment]
I would like to note that there's a working quasiquote with macros in [anise's core/macros.fnl](https://github.com/bb010g/anise/blob/master/core/macros.fnl). The syntax is pretty meh, though. [anise's macros.fnl](https://github.com/bb010g/anise/blob/master/macros.fnl) has a fair amount of usages.

**Comment by bakpakin** [comment]
> I would like to note that there's a working quasiquote with macros in anise's core/macros.fnl. The syntax is pretty meh, though. anise's macros.fnl has a fair amount of usages.

I think the anise macros look really good, is there any possibility we could merge these into the core? I think adding more macros to the core library would really help improve the usability of the language. We could also modify the parser to provide better integration for backquoting and/or normal quoting.

**Comment by technomancy** [comment]
Fixed in #125
